File: android/ac-templating/src/main/kotlin/com/microsoft/adaptivecards/templating/DataContext.kt
package com.microsoft.adaptivecards.templating

/**
 * Represents a data context for template evaluation with support for nested contexts
 */
class DataContext(
    /**
     * The current data value
     */
    val data: Any?,
    
    /**
     * The root data value (top-level context)
     */
    val root: Any? = data,
    
    /**
     * The current index when iterating over arrays
     */
    val index: Int? = null,
    
    /**
     * Parent context for nested scopes
     */
    val parent: DataContext? = null
) {
    /**
     * Resolve a property path in the current context
     * @param path Property path (e.g., "user.name", "$root.title", "$index")
     * @return The resolved value or null if not found
     */
    fun resolve(path: String): Any? {
        // Handle special variables
        when (path) {
            "\$data" -> return data
            "\$root" -> return root
            "\$index" -> return index
        }
        
        // Handle path starting with $root
        if (path.startsWith("\$root.")) {
            val remainingPath = path.substring(6) // Remove "$root."
            return resolvePath(remainingPath, root)
        }
        
        // Handle path starting with $data
        if (path.startsWith("\$data.")) {
            val remainingPath = path.substring(6) // Remove "$data."
            return resolvePath(remainingPath, data)
        }
        
        // Regular property path - resolve from current data
        return resolvePath(path, data)
    }
    
    /**
     * Resolve a property path in a given object
     * @param path Property path (e.g., "user.name")
     * @param obj The object to resolve from
     * @return The resolved value or null
     */
    private fun resolvePath(path: String, obj: Any?): Any? {
        if (obj == null) return null
        
        val components = path.split(".")
        var current: Any? = obj
        
        for (component in components) {
            if (current == null) return null
            
            current = when (current) {
                is Map<*, *> -> {
                    @Suppress("UNCHECKED_CAST")
                    (current as? Map<String, Any?>)?.get(component)
                }
                is List<*> -> {
                    val index = component.toIntOrNull()
                    if (index != null && index >= 0 && index < current.size) {
                        current[index]
                    } else null
                }
                else -> {
                    // Try to use reflection for custom objects
                    try {
                        val field = current.javaClass.getDeclaredField(component)
                        field.isAccessible = true
                        field.get(current)
                    } catch (e: Exception) {
                        null
                    }
                }
            }
        }
        
        return current
    }
    
    /**
     * Create a child context for iteration
     * @param itemData The item data
     * @param itemIndex The iteration index
     * @return A new child context
     */
    fun createChild(itemData: Any?, itemIndex: Int): DataContext {
        return DataContext(
            data = itemData,
            root = root,
            index = itemIndex,
            parent = this
        )
    }
}

File: ios/Sources/ACTemplating/DataContext.swift
import Foundation

/// Represents a data context for template evaluation with support for nested contexts
public final class DataContext {
    /// The current data value
    public let data: Any?
    
    /// The root data value (top-level context)
    public let root: Any?
    
    /// The current index when iterating over arrays
    public let index: Int?
    
    /// Parent context for nested scopes
    public weak var parent: DataContext?
    
    /// Initialize a root data context
    /// - Parameter data: The root data object
    public init(data: Any?) {
        self.data = data
        self.root = data
        self.index = nil
        self.parent = nil
    }
    
    /// Initialize a nested data context
    /// - Parameters:
    ///   - data: The current scope data
    ///   - root: The root data object
    ///   - index: The current iteration index
    ///   - parent: The parent context
    public init(data: Any?, root: Any?, index: Int?, parent: DataContext?) {
        self.data = data
        self.root = root
        self.index = index
        self.parent = parent
    }
    
    /// Resolve a property path in the current context
    /// - Parameter path: Property path (e.g., "user.name", "$root.title", "$index")
    /// - Returns: The resolved value or nil if not found
    public func resolve(path: String) -> Any? {
        // Handle special variables
        if path == "$data" {
            return data
        } else if path == "$root" {
            return root
        } else if path == "$index" {
            return index
        }
        
        // Handle path starting with $root
        if path.hasPrefix("$root.") {
            let remainingPath = String(path.dropFirst(6)) // Remove "$root."
            return resolvePath(remainingPath, in: root)
        }
        
        // Handle path starting with $data
        if path.hasPrefix("$data.") {
            let remainingPath = String(path.dropFirst(6)) // Remove "$data."
            return resolvePath(remainingPath, in: data)
        }
        
        // Regular property path - resolve from current data
        return resolvePath(path, in: data)
    }
    
    /// Resolve a property path in a given object
    /// - Parameters:
    ///   - path: Property path (e.g., "user.name")
    ///   - object: The object to resolve from
    /// - Returns: The resolved value or nil
    private func resolvePath(_ path: String, in object: Any?) -> Any? {
        guard let object = object else { return nil }
        
        let components = path.split(separator: ".").map(String.init)
        var current: Any? = object
        
        for component in components {
            guard let currentValue = current else { return nil }
            
            if let dict = currentValue as? [String: Any] {
                current = dict[component]
            } else if let array = currentValue as? [Any], let index = Int(component) {
                guard index >= 0 && index < array.count else { return nil }
                current = array[index]
            } else {
                // Try to use reflection for custom objects
                let mirror = Mirror(reflecting: currentValue)
                current = mirror.children.first { $0.label == component }?.value
            }
        }
        
        return current
    }
    
    /// Create a child context for iteration
    /// - Parameters:
    ///   - data: The item data
    ///   - index: The iteration index
    /// - Returns: A new child context
    public func createChild(data: Any?, index: Int) -> DataContext {
        return DataContext(data: data, root: root, index: index, parent: self)
    }
}
